### ISSUE TITLE
aria-label is ignored unless all the child content is hidden. 

### ISSUE THREAD

**Issue by andrii-morozov** [nvaccess/nvda#7360]
### Steps to reproduce:
[Example](https://codepen.io/anon/pen/vZroMb)
Focus the first div. Notice that aria-label is ignored only button content is read.
****

### Expected behavior:
I would expect the same behavior as for JAWS and Narrator. The aria-label is read in the first case. 

### Actual behavior:
Aria-label is ignored.

### System configuration:
Windows 10. Firefox 54.0 (32-bit). NVDA 2017.1 Portable

### Other information
I found the thread that you closed with similar issue but you fixed this just for buttons/links. Though the main point here is that I would prefer NVDA is doing the same things as the other screen readers instead of trying to do something different even though it seems smart.
The documentation doesn't have statements that [aria-label](https://www.w3.org/TR/wai-aria/states_and_properties#aria-label) can be ignored

**Comment by jcsteh** [comment]
It's true that the spec doesn't say aria-label can be ignored. However, it also only talks about labels. It does not talk about what should be reported as content. Content and labels are not the same. By way of example, if we treated the div in your example the same way we do links, we would override the button altogether, since we override the content with the label in the case of buttons, links, etc.

As for Narrator, it does report the label of the div when tabbing, but it completely skips the label when using scan mode. This further emphasises the point: treatment of aria-label is nowhere near as simple as it seems.

See my [blog post on aria-label](http://blog.jantrid.net/2015/12/woe-aria-surprisingly-but-ridiculously.html) for some further background on the complexity here.

What are you actually trying to achieve here? When the div gets focus, NVDA reads the content of the div, since that's normally what you want when a container gets focus. What is the real world use case here?

**Comment by andrii-morozov** [comment]
So we have pretty complicated blocks. And as you know modern web apps are not as simple as websites.  We have chart containers which if the NVDA starts to read it would be pretty confusing. Because it woudn't make a lot of sense for users. So we created aria-label attribute which creator can use to describe the chart. Other than that blind user can click see data which will navigate him to simple html table which can be navigated and read by virtual cursor. 

I have multiple points why we should read aria-label and ignore the content inside: 
         1. It follows the same pattern as JAWS and Narrator by default. Overall we don't want one more library to make screen readers more consistent as we had with JQuery.
         2. Because aria-label is custom attribute and I am as creator specified it I expect it will be used by screen reader.
         3. Modern Web Apps are not as simple as just web site as 10 years ago when virtual cursor was enough to make web site accessible. We have 10th of live areas we can have more than 10 000 elements with lots of content. So we want to give user better experience of navigating through content.

**Comment by andrii-morozov** [comment]
Other thing I was considering is to make all internal content aria-hidden though it not so simple because the tabular data is also part of this container. It means if the user will go to see data mode I will need to dynamically remove the aria-hidden attribute. Also the container has the control buttons which I don't want to be read by default. Though user can still access it through hot keys. 

Overall we want more control on what screen reader is reading.

**Comment by jcsteh** [comment]
You say you want us to render aria-label and ignore the content inside. If we do this, the user won't be able to get to the button because the button is *inside* the labelled div. Are you saying you don't want the button to be rendered? You either want us to override the elements inside or not; it can't be "override some content except buttons".

**Comment by jcsteh** [comment]
Also, as I've already pointed out, Narrator does not render the div in browse modee, so your arguments concerning Narrator are not quite correct. Furthermore, users do not simply navigate with the tab key; they primarily use browse mode.

**Comment by andrii-morozov** [comment]
The problem is that browse mode doesn't work well always. If you go to excel online they actually doing pretty tricky things in order the user had good experience. They are having specific aria-live which changes whenever you tab/arrow key to different cell. They don't even focus the cell as it is. So if the user will use browse mode I am sure the experience will be awful because the cells are not table and don't provide solid navigation experience.

However I understand your point. Maybe I need to come with different design and do something as tricky as Excel does. The main problem for me is that I started to test NVDA a bit late and JAWS and Narrator worked well for me.

**Comment by jcsteh** [comment]
Something like Excel (or Google Sheets) is a good example of where browse mode is totally irrelevant. For that reason, it's also one of the few examples where role="application" is a good idea. Note that in focus mode (and role="application", where browse mode isn't even considered), NVDA will read your aria-labels when you focus just fine. The reason it doesn't in your example is that you're dealing with browse mode.

**Comment by andrii-morozov** [comment]
Thanks for help. Will see what I can do. Ideally looking towards role="application"  at least for charts.
